### ISSUE TITLE
Tagging sites for the first time is cumbersome

### ISSUE THREAD

**Issue by MarcusRiemer** [mozilla/multi-account-containers#719]
If I understand the UI correctly, categorizing a site that I have already opened in a tab requires me to:

1. Open an empty tab in the correct container
2. Copy the URL from the existing tab to the correctly "containered" tab
3. Click the "Container" icon
4. Select the "Always open in this container"-checkbox

If this is working as intended: It would be great to have a less tedious way to do this.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/MAC-580)

**Comment by jonathanKingston** [comment]
Confirm, this is exactly the flow to assigning sites.

Currently we are pretty busy at the moment however it is something I wanted to look at, perhaps with a full page management interface of all URLs you ever assigned and an easy way to just type in new sites.

**Comment by jonathanKingston** [comment]
Also, thanks for raising this. I often store my ideas in my head which isn't great.

**Comment by MarcusRiemer** [comment]
From my POV simply having a keyboard shortcut would be a viable workaround for powerusers while requiring not too much (tm) work.

**Comment by ArchangeGabriel** [comment]
@jonathanKingston A management interface would be of no help, if you already have the URL at hand you could just as well open it in the right container.

The only thing that would work here AFAICT is being able to reload an existing tab into a different container. But per https://github.com/mozilla/testpilot-containers/wiki/Moving-between-containers, this is not going to happen apparently.

**Comment by jonathanKingston** [comment]
> The only thing that would work here AFAICT is being able to reload an existing tab into a different container.

Yeah not happening, however we have discussed initial setup being able to pull tabs into a new container. This however requires clean up of cookies which is pretty similar in work unless you as a user just wanted a quick way to move urls and be logged out.

> A management interface would be of no help

Not entirely, seeing all container assignments in one place would be super useful for advanced users with lots of assignments. Also someone else requested a JSON or CSV upload or something, that really won't fit into the current setup so would be useful there (imagine having shareable lists of Social, banking etc).

> From my POV simply having a keyboard shortcut

Kinda disagree, finding a shortcut to work to open the management interface as it is was hard enough (I believe it still doesn't work on Mac either). A user can currently however tab through the interface after opening it. Designating one just for assignment would be a waste of a limited set of shortcuts IMO.

**Comment by m-j-r** [comment]
>A management interface would be of no help
I disagree completely, as I imagine that such interface would allow for backing up/restoring configurations (so that you only need to assign sites once), importing curated containers, etc.

**Comment by Caemor** [comment]
Maybe sth similar to the context plus addon (https://addons.mozilla.org/de/firefox/addon/context-plus/) to just rightlick on an already open tab to be able to move it to a different container would be a great help for first tagging.

**Comment by ArchangeGabriel** [comment]
The “advanced” management interface exists, it’s here: `~/.mozilla/firefox/<profile-id>/browser-extension-data/@testpilot-containers/storage.js`. ;p

**Comment by madb1lly** [comment]
I'd like to be able to manually edit which sites open in a container. E.g. I've created a new tab for accounts.google.com, but maybe I want * .google. * to open in that tab - manually editing it would be much quicker than having to select the container each time I visited one of that group of sites.

**Comment by maxnordlund** [comment]
This would be pretty sweet to have. Some sort of table with editable patterns. This would also naturally support CSV import/export in the future.

By grouping the patterns inside the table you get a bit nicer UX for not too much effort.

```html
<table>
    <thead>
      <tr><th>Pattern</th></tr>
    </thead>
    <tbody>
      <tr><th>Personal</th></tr>
      <tr><td><input type="text" value="google.com" /></td></tr>
      <tr><td><input type="text" value="twitter.com" /></td></tr>
      <tr><th>Work</th></tr>
      <tr><td><input type="text" value="github.com" /></td></tr>
    </tbody>
</table>
```

**Comment by sixpointzero** [comment]
Are there any plans to add a simple drop down list of containers where "Always open in Default"  is written? I don't think the discussion in this thread went where @MarcusRiemer intended.

If I have opened a website in a default non-container tab, I'm unable to assign it to a container. Instead, I have to follow the tedious flow of opening that site again in the desired container and then carrying out the assignment. I'm aware that this add-on won't implement the "moving between containers" functionality but this case isn't the same.

**Comment by chuckadams** [comment]
At the very least, moving or copying the "open in container" menu from tabs to the container button would do a lot for the UX.  In fact, it's exactly the behavior I expected from the button menu in the first place, but I was disappointed to find that all it does is open a new blank tab in the given container.

**Comment by zeorin** [comment]
The fact that [an extension exists that does exactly this](https://addons.mozilla.org/en-GB/firefox/addon/containerise/) is pretty strong proof it's a widely desired feature. Be nice if it was official Mozilla, though.

**Comment by mmoraschini** [comment]
I'm also upvoting this. Would be nice to specify for each container a list of sites that automatically open in the given container and all those that are not in any list in "no container" or a default one.

**Comment by TriMoon** [comment]
No idea how they count the votes, but i have added a :+1: to the OP **and** mentioned this issue in my own report(s) **and** posted a comment here **and** pinged @groovecoder  now :rofl: 
Comon guys just do it...

**Comment by luger19** [comment]
We are still waiting...

**Comment by maxnordlund** [comment]
Please, such comments aren't helping. Just upvote the issue and move on. It's like a bad boss/product owner hover over your shoulders asking "When will it be done?" for the 100th time.

**Comment by luger19** [comment]
Well, @MarcusRiemer opened this issue back on Aug 10, 2017. It's just a little annoying that nothing's happened yet... (no offence)

**Comment by ArloL** [comment]
Hi @luger10,

thanks for showing an interest in this issue and that you even commented. That is awesome! Most people do not make their voice heard. You apparently like the extension so much that you searched for an existing issue and found it. That is great! Most people do not try and file bugs or feature requests - they just uninstall or use what is there and are slightly annoyed day by day. Those are some great qualities you presented.

You probably came here being slightly annoyed every time you had to do these steps manually and now you are majorly annoyed because "how could someone not just implement this? the issue has been open since 2017". You would like to somehow get your way and that someone just implement this and you have chosen to comment with "We are still waiting...".

Do you think it's a possibility that your comments are actively preventing you from reaching your own goal?

Please do think of the maintainers and that most of them (maybe even all) are doing the work in their free time. My recommendation would be to write something like this:

> I have been happily using MAC for two years now. But I just had to setup a new computer and had to do all this setup manually again and remembered this issue. Is there anything preventing work on this? Please let me know if I can help. Would love to see this feature 😍

Is it perfect? No. Will it achieve your goal? Maybe no. Maybe yes. Will it put unnecessary unwarranted pressure on a maintainer? Absolutely not.

Hope your safe and that next time you might think about how you come across to people on the other side of the screen. We are all just humans after all 🙂

**Comment by maxnordlund** [comment]
Well said @ArloL, and I think a good point is also that is you want something, why not try to contribute it yourself? I'm sure the people around here are more then willing to help if you ask nicely 😄

**Comment by TriMoon** [comment]
@ArloL @maxnordlund @luger19 
Being irritated/annoyed and being able to contribute are two **entirely different** things.
If people who are actually able to contribute (aka developers) get irritated/annoyed enough by the irritation/annoyance of common users they will be driven more to **finally implement** something to eliminate that irritation/annoyance.
So people voicing their irritation is IMHO **perfectly fine** and should be encouraged instead of frowned upon.

Most of us are adequate in coding in many languages, but are unable to contribute because of many different reasons. _(among which most probably the fact that we are over our ears in other stuff already.)_
So lets not throw the ball around to others or even bounce it back to people who voice their feelings, if you are not able to do it yourself...
Because THAT is just weak personality...

Bottom line:
Agree or disagree with a reply, but do not discourage them to voice their feels.

As you can notice i am not discouraging your pov's on @luger19's reply but i am disagreeing with it. With the intend to improve your social communication styles...

**Comment by maxnordlund** [comment]
>  Being irritated/annoyed and being able to contribute are two *entirely different* things.

Yes they are. Which is kinda the point I was making. If you sound aggressive, or even yelling, does *not* make you want to work on something in your *free time*. _(Emphasis inspired by @TriMoon™️)_

> If people who are actually able to contribute (aka developers)

I'm not sure if you noticed but I'm fairly sure this platform is pretty developer heavy 😉. But seriously, this is the development platform for MAC. The place where code review is done. I think it's fair to assume people around here are developers in some capacity.

> get irritated/annoyed enough by the irritation/annoyance of **common users** 

Again, there are no "common" users vs developers. We're all users, and most if not all here on GitHub are also developers.

Yes, it's about scratching

> Most of us are adequate in coding in many languages, but are unable to contribute because of many different reasons. _(among which most probably the fact that we are over our ears in other stuff already.)_

This is not a valid reason. You expect someone else to do your work for you just because *you* don't have time? Yet you correctly assert that many don't have extra capacity to contribute.

> they will be driven more to *finally implement* something to eliminate that irritation/annoyance.

I think this hit it on the nail. You scratch an itch, and if it bugs you enough, you get it done yourselves. Hopefully you also share your work so that others can benefit from it as well. In the best of worlds we would all do this from the kindness of our heart, but I'll settle for collaboration in any shape and form.

If you need help, ask. If you don't know the language, it's easy enough to pick up if you already have experience in an imperative, strictly evaluated language. Bonus points if it's dynamically typed (JS is also weakly typed, but that's not too common).

_Yes, there are some foot-guns in JS, but there are those in all languages. Just stay away from `with` statement and messing with builtin prototypes and your good._

**Comment by madb1lly** [comment]
Hi all,

I know we're way off the purpose of this issue's discussion now, yet I have to point out that _no_ one can assume that most people on Github are developers. The Firefox add-ons page for multi-account containers sends people to this repo's issues page for "Support site". Therefore there will probably be more "common users" here than developers, hence it's not reasonable to assume that most people here are developers.
![image](https://user-images.githubusercontent.com/25123631/77291446-91084800-6cde-11ea-8814-e3751f1e799e.png)

This also means that lots of people coming here might not know the established issues/bugs discussion etiquette that is well-established in Mozilla and the wider FLOSS community. So, let's try and give people the benefit of the doubt, help them learn and when we're finished we'll all navigate away from this page in a much better mood 😊

Please excuse me if this comes across as preaching, I tried to write it in a non-patronising way but it might not have worked.

Cheers 🙂

**Comment by maxnordlund** [comment]
I thought it was good. Didn't know about the support site part. Interesting indeed.

**Comment by TriMoon** [comment]
> @maxnordlund 
> Didn't know about the support site part.

Besides the point that **everyone** who uses some software is included in the "general user" category.
And that only contributors with actual code are considered "developers" for any software.
You can be a developer for software "A", but still be a general user for software "B" despite your ability to become a developer for software "B" also.

When it comes to the point of "doing it yourself":
- When developer "X" makes/contributes to software "A" and common user "Y" finds a bug or missing feature or usage problem in the software and reports it to the developers of software "A";
   - Then developer "X" is **NOT** entitled to say "do it yourself" to common user "Y", because it is not that users responsibility to do so, otherwise persons "X" and "Y" will immediately change positions per definition.

That said you should stop viewing common users as possible developers because of your personal inabilities...
If every developer would think like you do then there would be no FOSS nor commercial software at all, at least no usable ones...
Commercial software also depends on FOSS in one way or the other, there is no company that has software that is coded by a single person <s>except maybe the first electronic accumulator</s> _(no that also used info from others...)_

Anyway, recompile your terminology and opinions and...
# **Lets return to the original topic..**

**Comment by maxnordlund** [comment]
OK, so first off please tone it down a notch. Then let me try to condense it as much as I can. Full disclosure: I haven't committed anything in this repo, but if I find something I can fix I most definitely will make a PR.

---

Reporting a bug is great; feeling frustrated that something doesn't work or that a bug has been open for years is totally fine. It's something I do myself all the time. What I don't, or try my hardest not to, do is be aggressive about it. Passive or otherwise.

The written word often comes across as more harsh then one might think, and it's a real skill to communicate clearly. That's why we have copywriters who's only job is to write such crystal clear text. In this context this means you and me should strive to write as clear and neutral/positive as possible. This is hard.

Then there's the other part of this discussion. Whether a user reporting an issue should be encouraged to contribute a fix or not.

I believe so, yes. Because how else would you recruit new maintainers for an FLOSS project? By encouraging and guiding users to become developers, to use your terms, you keep the project going. To be successful it's paramount to also provide help and guidance. Personally only can help teach JS and some basic web extension stuff, but as stated above I don't know the code base. I am, however, sure that the maintainers here are willing to help onboard new developers if asked nicely.

---

_A small side, trying to shut down discussions with an impressively large text is not neutral or positive. Just sayin' 😉_

**Comment by AdKiller** [comment]
A way to type global wildcard like (containerise addon) is sorely missing.

For example *.bandcamp.com to open any band site in Bandcamp container.

**Comment by grahamperrin** [comment]
https://github.com/mozilla/multi-account-containers/issues/719#issuecomment-442110216

> … "open in container" … button …

The button now includes two relevant menu options, here's the second: 

![image](https://user-images.githubusercontent.com/192271/115924575-d4039480-a477-11eb-9a09-9cc3ed375eb1.png)

----

## Opening post

> ## Tagging sites for the first time is cumbersome

<https://github.com/mozilla/multi-account-containers/issues/719#issue-249297501>: 

> 1. Open an empty tab in the correct container
> 2. Copy the URL from the existing tab to the correctly "containered" tab
> 3. Click the "Container" icon
> 4. Select the "Always open in this container"-checkbox
> 
> … have a less tedious way …

https://github.com/mozilla/multi-account-containers/issues/719#issuecomment-321760537

> … keyboard shortcut would be a viable workaround for powerusers …

@MarcusRiemer please: given the screenshot above plus shortcut for **Open containers panel**, do things now feel less cumbersome/tedious? 

If not, there's a three-click method. Between the first and second clicks: 

![image](https://user-images.githubusercontent.com/192271/115927512-71f95e00-a47c-11eb-831b-ba57b75e668b.png)

**Comment by kurahaupo** [comment]
Question: where are the containers's domains stored?
I would be happy enough editing a text file (or json) or using sqlite to modify a database.

**Comment by adelsz** [comment]
> Question: where are the containers' domains stored? I would be happy enough editing a text file (or json) or using sqlite to modify a database.

I had to make this snippet to make adding domains easier (especially redirecting ones):
https://gist.github.com/adelsz/663393c06adcceb5ec564f22e2c71c2a
Note that you need to run it through the extension background script console.

**Comment by achernyakevich-sc** [comment]
This issue is quite old. Since it was created passed a lot of time and many things were changed in the world. :)

MAC already provided several ways to tag some site to be "Always open in the container". From other side users' expectation even become wider. Additionally we need to notice that MAC developers are quite busy and try to concentrate on the core add-on's functionality but provide great possibility for integration for other developers.

For many cases described here I would recommend to use [Containerise](https://addons.mozilla.org/en-US/firefox/addon/containerise/) add-on. Easy to use and bring much more power for sites tagging than MAC itself.

Maybe even it has a reason to close this issue? :)

**Comment by kurahaupo** [comment]
@achernyakevich-sc Yes it provides several ways of tagging sites, but as yet I can see no way to add a domain of any URI that elicits an HTTP redirection (301, 302, 303, 307, 308) response, nor of adding the domains for components (especially css & js, though frames, images, and other media are also problematic).

Or have I missed something?

Also, "MAC developers are busy" doesn't explain why several PRs that would improve this (submitted by various people including myself) have simply been ignored.

**Comment by achernyakevich-sc** [comment]
> @achernyakevich-sc Yes it provides several ways of tagging sites, but as yet I can see no way to add a domain of any URI that elicits an HTTP redirection (301, 302, 303, 307, 308) response, nor of adding the domains for components (especially css & js, though frames, images, and other media are also problematic).

It looks I got your wish - you would like to be able to load a page's  resources in containerised mode. But MAC looks designed different way. It isolates the loaded page completely. Be frankly I don't see even a reason to load a page's resources in other cotnainers as it will easily break page contents (as for server these will be loaded form different browser, etc.)

> Or have I missed something?
> 
> Also, "MAC developers are busy" doesn't explain why several PRs that would improve this (submitted by various people including myself) have simply been ignored.

I'm not MAC developer and can't say anything about "several PRs that would improve this (submitted by various people including myself) have simply been ignored". This is just an Open Sources. If you would like you could fork this repository and start to maintain own version of this add-on. Being more actively developed it could become very attractive for users and be installed more times than this one.

**Comment by dannycolin** [comment]
"From my POV simply having a keyboard shortcut would be a viable workaround for powerusers while requiring not too much (tm) work." -- OP

Since OP considered that keyboard shortcuts were a viable workaround and that this feature is now available in MAC, I will close this issue. 

I'd also want to point out that #691 and #1670  should be the issue to upvote for those who proposed regex/pattern matching and manually editing the site list as a solution. 

Thanks all.

**Comment by xanoni** [comment]
In case people haven't seen it in the other tickets, my workaround has been to somehow get the HTTP server behind the subdomain to generate a `404` response (instead of redirecting me). Then you can add the subdomain via the plugin.

See https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-1416926881
